bdk_cli_lib/src/lib.rs

```diff
@@ -478,42 +502,89 @@ where
 {
     match command {
         // TODO: Make these functions return stuffs
-        Commands::Address { addr_cmd } => {
-            run_address_cmd(tracker, store, addr_cmd, network)?;
-        }
+        Commands::Address { addr_cmd } => run_address_cmd(&tracker, &store, addr_cmd, network),
         Commands::Balance => {
             run_balance_cmd(&tracker);
+            Ok(())
         }
         Commands::TxOut { txout_cmd } => {
-            run_txo_cmd(txout_cmd, tracker, network);
+            run_txo_cmd(txout_cmd, &tracker, network);
+            Ok(())
         }
         Commands::Send {
             value,
             address,
             coin_select,
         } => {
-            let transaction = create_tx(value, address, coin_select, tracker, &keymap)?;
-            let changeset = tracker.insert_tx(transaction.clone(), P::unconfirmed())?;
-            client.broadcast(&transaction)?;
-            // We only want to store the changeset if we actually successfully broadcasted because
-            // it will increase the derivation index of the internal keychain.
-            store.set_derivation_indices(tracker.txout_index.derivation_indices())?;
-            store.append_changeset(&changeset)?;
-            println!("Broadcasted Tx : {}", transaction.txid());
+            let (transaction, change_index) = {
+                // take mutable ref to construct tx -- it is only open for a short time while building it.
+                let tracker = &mut *tracker.lock().unwrap();
+                let (transaction, change_info) =
+                    create_tx(value, address, coin_select, tracker, &keymap)?;
+
+                if let Some((change_derivation_changes, (change_keychain, index))) = change_info {
+                    // We must first persist to disk the fact that we've got a new address from the
+                    // change keychain so future scans will find the tx we're about to broadcast.
+                    // If we're unable to persist this then we don't want to broadcast.
+                    let store = &mut *store.lock().unwrap();
+                    store.append_changeset(&change_derivation_changes.into())?;
+
+                    // We don't want other callers/threads to use this address while we're using it
+                    // but we also don't want to scan the tx we just created because it's not
+                    // technically in the blockchain yet.
+                    tracker.txout_index.mark_used(&change_keychain, index);
+                    (transaction, Some((change_keychain, index)))
+                } else {
+                    (transaction, None)
+                }
+            };
+
+            match client.broadcast(&transaction) {
+                Ok(_) => {
+                    println!("Broadcasted Tx : {}", transaction.txid());
+                    let mut tracker = tracker.lock().unwrap();
+                    match tracker.insert_tx(transaction.clone(), P::unconfirmed()) {
```

Comment:
I think it would make more sense to try insert it into `tracker` prior to broadcasting and directly after the transaction is created. This way, we can check the created transaction against our `ChainGraph` rules.

If broadcast fails because the transaction cannot be accepted into the mempool, we can have a method to remove unconfirmed transaction from `SparseChain`/`ChainGraph`, and signal to the user that we need a sync or scan.

If broadcast fails because of network issues, we can retry until a timeout.

Reply:
> I think it would make more sense to try insert it into `tracker` prior to broadcasting and directly after the transaction is created. This way, we can check the created transaction against our `ChainGraph` rules.

Isn't this what `insert_tx_preview` is for?

We can't insert it before unless we are either (i) willing to save it to db at that time OR (ii) keep a lock on the db while we broadcast.  (i) is bad because then the change address will always be marked as used (txout index will mark it as such when it sees it come on from DB) and (ii) is bad because it keeps a potentially long lock on the db.

Reply:
> Isn't this what insert_tx_preview is for?

Reply:
I still think (i) is ideal. We would want to keep the change address marked as used since we have already potentially exposed it (via broadcast).

Reply:
but what if we know we have not exposed it? i.e. broadcast failed in a way that does not expose it.